pkg/service/auth_pipeline.go

```diff
@@ -427,42 +465,33 @@ func (pipeline *AuthPipeline) Evaluate() auth.AuthResult {
 		evaluateFunc := func() {
 			// phase 1: identity verification
 			if resp := pipeline.evaluateIdentityConfigs(); !resp.Success() {
-				code := rpc.UNAUTHENTICATED
-				pipeline.reportStatusMetric(code)
-				authResult <- pipeline.customizeDenyWith(auth.AuthResult{
-					Code:    code,
-					Message: resp.GetErrorMessage(),
-					Headers: pipeline.AuthConfig.GetChallengeHeaders(),
-				}, pipeline.AuthConfig.Unauthenticated)
-				return
+				result.Code = rpc.UNAUTHENTICATED
+				result.Message = resp.GetErrorMessage()
+				result.Headers = pipeline.AuthConfig.GetChallengeHeaders()
+				result = pipeline.customizeDenyWith(result, pipeline.AuthConfig.Unauthenticated)
+			} else {
+				// phase 2: external metadata
+				pipeline.evaluateMetadataConfigs()
+
+				// phase 3: policy enforcement (authorization)
+				if resp := pipeline.evaluateAuthorizationConfigs(); !resp.Success() {
+					result.Code = rpc.PERMISSION_DENIED
+					result.Message = resp.GetErrorMessage()
+					result = pipeline.customizeDenyWith(result, pipeline.AuthConfig.Unauthorized)
+				} else {
+					// phase 4: response
+					pipeline.evaluateResponseConfigs()
+					responseHeaders, responseMetadata := evaluators.WrapResponses(pipeline.Response)
+					result.Headers = []map[string]string{responseHeaders}
+					result.Metadata = responseMetadata
+				}
 			}
 
-			// phase 2: external metadata
-			pipeline.evaluateMetadataConfigs()
-
-			// phase 3: policy enforcement (authorization)
-			if resp := pipeline.evaluateAuthorizationConfigs(); !resp.Success() {
-				code := rpc.PERMISSION_DENIED
-				pipeline.reportStatusMetric(code)
-				authResult <- pipeline.customizeDenyWith(auth.AuthResult{
-					Code:    code,
-					Message: resp.GetErrorMessage(),
-				}, pipeline.AuthConfig.Unauthorized)
-				return
-			}
+			// phase 5: notify
+			pipeline.evaluateNotifyConfigs()
```

Comment:
> Could we move this to be async? I realize this might be a bigger change than it sounds

It is not a big change and with a cost worth paying IMO. Being this truly about notification, I think it's totally reasonable that the responses of the `notify` phase are not added back to the authorization JSON. If people can't build other steps of notification on top of previous notification results obtained in the same pipeline, and that is the price to pay for lower latency, I think it's a fair price to pay.

That said, the reason why I haven't made it async before is because I wanted to double check on the implications (risks?), given the (bad) way we currently manage the memory. The AuthPipeline does not know about implementation details in the evaluators, yet it passes self by reference so the evaluators can access (read), e.g., the authorization JSON. Maybe that's fine, but I'd like to run some tests before committing to anything along those lines.